### PR TITLE
[processing] some changes to make it easier to handle invalid features

### PR DESCRIPTION
**Pull request by volaya**
This moves the ideas implemented in https://github.com/qgis/QGIS/commit/143cfab1047f13db7654a4f241c1716d0d45c238 to Processing core, so they can easily be adapted by new algorithms.

### CODE REVIEW

**Comment by nyalldawson** [comment]
What's the future plans for this? Will this just be introduced to select algorithms?

**Comment by volaya** [comment]
@nyalldawson For no, those are just some handy methods, to make it easier to extend the behaviour that @m-kuhn added to one of the algorithms, so all work in the same way without having to repeat the code.

Eventually, i guess we should add that to most of the algorithms. I would actually propose having that done automatically, so for all algs using vector layers, the corresponding "ignore invalid geometries" is added by default (the code of the alg should be adapted, though, to use the value of that parameter)

Going further, i am thinking of not just adding that, but also some mechanism to fix invalid geoms, which would appear as an option as well in all vector algs. I have been discussing it with @gioman, maybe to have it based on the lwgeom library (which would have to be added to QGIS builds...)

But anyway, that's another thing. For now, i would like to have this handy methods in there and start adapting the most basic algorithms (the computational geometry ones, like intersect, difference, etc, that come from ftools) to have a parameter to ignore invalid geoms

**Comment by nyalldawson** [comment]
So, my concern is the performance impact this will have on processing. Forcing a validity check for every use hurts users who look after their data.  It also is quite wasteful with multiple step models.  For this reason I also disagree with the original commit.

I think we should just leave the current behaviour, and if people feed in bad geometry they get unexpected answers. Garbage in, garbage out. It's how postgis (and probably other tools) work too.

If anything this should be a global processing option which defaults to off.

@m-kuhn as author of the original change what do you think?

**Comment by gioman** [comment]
> So, my concern is the performance impact this will have on processing. Forcing a validity check for every use hurts users who look after their data. It also is quite wasteful with multiple step models. For this reason I also disagree with the original commit.
> 
> I think we should just leave the current behaviour, and if people feed in bad geometry they get unexpected answers. Garbage in, garbage out. It's how postgis (and probably other tools) work too.

@nyalldawson 

Hi Nyall, this hurts a lot common users and hurts (since a long now) QGIS reputation. it is expected that at least the tools will return a meaningful message warning that input is bad, and at the moment this is frequently only shown in QGIS log, that as we know user will not check.

Moreover an easy way (I underline "easy") to fix geometries is expected as a tool and also as option within other geoprocessing tools. Enabled by default or not is no problem for me, but it it should exist. 

Just a note:
the overhead of cleaning beforehand geometries with (example) liblwgeom would be negligible in the vast majority if cases. It anyway much faster than using v.clean for example, as we know there is the overhead of importing/exporting the data in the temp GRASS mapset. As far as I know there are no other ways to clean geometries in processing.

so a big +1 for me (and as far as I can tell also from all the users I explained that knowing they are working with garbage (that frequently it is not their fault) is a good thing... and then failed to show them an easy way to solve this issue).

**Comment by m-kuhn** [comment]
I think a lot of tools right now print messages to the message log. So the performance impact is there but the warning is lost because it's not visible enough.

If there are concerns for performance, we could possibly move the code into the C++ iterator possibly combined with a QgsFeatureRequest flag.

I also think that it's ok to say garbage in, garbage out, but in this particular case, it's non-obvious that the data is garbage, even for professional, experienced users.

**Comment by gioman** [comment]
> I also think that it's ok to say garbage in, garbage out, but in this particular case, it's non-obvious that the data is garbage, even for professional, experienced users.

@m-kuhn 

It would only be ok if QGIS would return a meaningful message to users, something that does not (please do not bring here the QGIS log, is not real to ask common users to look and understand what's going there, if you think otherwise then maybe you don't know the vast majority of our users).

The result is that the output is considered wrong, and I think that it really is (an intersection produces a wrong result because of a single auto-intersection in a polygon among thousands of them? and what about if what I'm doing has legal implications? it remembers me when I had to compute areas at the time we knew that QGIS was screwing that computation, and had to guarantee before the law that the values were correct... just to say that sometime we do some real life work out here.)

QGIS geoprocessing tools should:

produce the right output, something that is not always granted even with correct inputs (see union, etc.)

warn clearly when input has problems

give a simple tool/option to clean geometries beforehand or within the operation that is needed to run

perform better: most of native qgis geoprocessing tools are unusable with large datasets, being all the alternatives available (ogr, grass, saga) faster, sometimes orders of magnitude

allow to be run in the backgroud

**Comment by mach0** [comment]
@gioman s list pretty much sums it up.
I sometimes have the same problems here and the longer we (QGIS) produce unreliable (or lets say not 100% correct) results, the more bad it is for the reputation. 
I already had friends that made me aware of some failures (see -> Union) but afterwards they told me to leave - and only come back to QGIS when the results are reliable.
being userfriendly is sometimes not enough.

**Comment by volaya** [comment]
@nyalldawson I understand your worries about the performance, but I think that the advantages in this case are more important than the disadvantages, as @mach0  and @gioman point out.

In any case, I think we can combine them all. Here is what I propose

- Not using a parameter in each algorithm but a Processing-wide option to define how to handle invalid features
- The option would have three possible configurations.:

1) Do not check for invalid features. It would not filter the feature iterator, so there will be no performance issues. This can be used by advanced users who know about their data and need no extra checks done by processing. This will be the fatest option.
2) Ignore invalid features. It uses the filter that is added in this PR, and just adds a message to the log in case of finding invalid features
3) Stop algorithm execution when valid features are found. It uses the filter that is added in this PR, and raises an exception. The most explicit of the 3 options, this one should be the default one.

Let me know what you think. I think this is a safe way of dealing with this.

I will work on adapting this PR to this approach, but let me know if you have suggestions.

**Comment by volaya** [comment]
PR is adapted. If you agree on this approach, I will edit the algorithms that deal themselves with invalid features (most of ftools one), so they rely on the core functionality added by this PR

**Comment by nyalldawson** [comment]
@volaya sounds good to me!

**Comment by m-kuhn** [comment]
sounds good to me as well!

I guess we handle NULL geometries in a second step. This will probably be something to spend some more time about when it's appropriate to dismiss NULL geometries, preserve them... depending on algorithms.

**Comment by pcav** [comment]
Or use MakeValid?

File: python/plugins/processing/tools/vector.py [hunk 1]
```diff
@@ -104,6 +108,27 @@ def __init__(self, layer, request):
...
+                if geom is None:
+                    ProcessingLog.addToLog(ProcessingLog.LOG_INFO,
+                                            self.tr('Feature with NULL geometry found.'))
+                elif not geom.isGeosValid():
+                    ProcessingLog.addToLog(ProcessingLog.LOG_ERROR,
+                                            self.tr('GEOS geoprocessing error: One or more input features have invalid geometry.'))
+                    if ignoreInvalid:
+                        return False
+                    else:
+                        raise GeoAlgorithmExecutionException(self.tr('Features with invalid geometries found. Please fix these errors or specify the "Ignore invalid input features" flag'))
```
Comment (DelazJ) [inline python/plugins/processing/tools/vector.py]: > or specify the "Ignore invalid input features" flag'

Or any hint on where the user can specify this option? :)

Comment (m-kuhn) [inline python/plugins/processing/tools/vector.py]: That would be the nicest solution of course, if that's going to happen my comment is pointless.

**Comment by gioman** [comment]
> Or use MakeValid?

yes, I really think we should rely on liblwgeom as a tool/option to give users an easy (and very good) way to clean their layers. V.clean should be left as a suggestion for the more advanced users, or as a secondary options in case the clean with makevalid does not work as expected,

**Comment by m-kuhn** [comment]
Is this the only (highly) interesting tool from lwgeom?
Trying to find out if it's better to copy this one function to our own QgsGeometry::makeValid() or make lwgeom a required dependency...

**Comment by nyalldawson** [comment]
@m-kuhn I think src/core/simplify/effectivearea.* was originally from lwgeom also, but it's been optimised to work with QgsGeometries directly. I'd be in favour of copying that single function if it's trivial, and then we can slowly optimise to our own internal geometry structure.

**Comment by m-kuhn** [comment]
I'll have a quick look

**Comment by nyalldawson** [comment]
@volaya Looks good, nice approach here. Could you add some unit tests covering this?

**Comment by fsdias** [comment]
@volaya @gioman @m-kuhn 

Since you think adding "Make valid" from lwgeom to correct invalid geometries is a good idea, I would suggest a 4th option:

4) Run "Make valid" on all input layers before intersection/union/dissolve etc. This option would of course have to be selected by the user on Processing - Options - General.

My reasoning is that invalid features are fairly common in public datasets. In some cases these geometries are very small and have no significant effect on the results of geoprocessing tools. In other cases the users are aware of them but just want to process a small part of a shapefile in order to produce a map. In these cases I think it would be useful to have QGIS correct invalid geometries on the fly.

I understand your concerns about reliability and correctness, but provided the user is made aware of what is being done I don't think this is a bad idea. For instance, a warning such as this would suffice in my opionion:

Warning: QGIS will run "make valid" on all the input shapefiles and will correct invalid geometries in case they exist. We recommend that you check the validity of the geometries beforehand.

**Comment by alexbruy** [comment]
Manually merged

**Comment by wonder-sk** [comment]
My two cents to the discussion:
- GEOS-based geoprocessing algs should _always_ fail loudly if there was GEOS exception when performing GEOS operation (i.e. returned geometry is null == isEmpty() returns True). The exceptions are typically a result of operation on invalid input geometry...
- for invalid geometry handling, I would suggest to have two options: let them go through (default) or skip them. It is IMHO important to avoid tests for geometry validity by default: not just performance-wise, but to also avoid surprises of not having all features processed.
- on failure of GEOS-based geoprocessing alg, we should suggest that invalid geometries are likely the reason of failure and that the user may want to fix the input data or let the alg skip invalid geometries.
- when skipping of invalid geometries is turned on, there should be a counter of number of skipped features so that user can find out that some features were skipped.
- in the future when we have MakeValid operation implemented for QgsGeometry, there could be third option to automatically fix invalid geometries. (note: in some cases even MakeValid may fail)

**Comment by m-kuhn** [comment]
I fully support the previous comments by @wonder-sk :+1: 

Can everyone involved with this topic agree on them?

@volaya , @alexbruy , @gioman , @nyalldawson , @fsdias

**Comment by nyalldawson** [comment]
Sounds good to me. Something else that occurred to me last night is that we need to way for individual algs to flag that ALL checks should be skipped. This should be used for algs which are not affected by the validity of geometries eg things like "extract by attribute" or table joins.

**Comment by gioman** [comment]
@m-kuhn @wonder-sk 

1) I think that we all agree on point one. This was asked when the title/description of 

http://hub.qgis.org/issues/9297

was changed to "ftools: warn users in clear, meaningful, visible way when geometries have errors (was : "ftools difference: wrong result")" 10 months ago.

2) 3) 4) 5) From a user point of view it is imperative that there is a *simple* way to handle/fix invalid geometries, both as standalone tool *and* as an *option* (no one asked to make it mandatory or not possible to disable) within standard geooprocessing tools (dissolve, union, etc.) in order to have the inputs fixed along the operation being run.

The overhead of fixing/cleaning input geometries with liblwgeom/st_makevalid is *neglectable* for the vast majority of real life cases, and when it adds some measurable time this is clearly compensated by not having huge headaches deriving by not needing to look for complicated (and much more time consuming) tools to clean what you call "garbage" (that by the way is the daily bread for who uses gis).

So... I agree with the above points, by "option 3" cannot wait anymore if we want to be useful also to people who do more geoprocessing rather than cartography.

Note: yes, in some cases even st_makevalid can fail, on the other hand it fixes the vast majority of the garbage without a glitch.

**Comment by m-kuhn** [comment]
> we need to way for individual algs to flag that ALL checks should be skipped

I think the question is should these checks be opt-in or opt-out.

What we could possibly do for this is

 * Have python bindings for `QgsGeometry::operation()` raise python exceptions if they fail.
 * Make the checks opt-in, no checks by default, introduce them when exceptions start popping up.

**Pull request closed without merging**